### PR TITLE
Allow setting custom error messages for preconditions

### PR DESCRIPTION
**Pull request by Joe4evr**

### CODE REVIEW

File: src/Discord.Net.Commands/Attributes/Preconditions/RequireBotPermissionAttribute.cs [hunk 1]
```diff
@@ -52,7 +53,7 @@ public override async Task<PreconditionResult> CheckPermissionsAsync(ICommandCon
                 if (guildUser == null)
                     return PreconditionResult.FromError("Command must be used in a guild channel");
```
Comment (k-boyle) [inline src/Discord.Net.Commands/Attributes/Preconditions/RequireBotPermissionAttribute.cs]: What about a second custom message for this one?

Comment (Joe4evr) [reply src/Discord.Net.Commands/Attributes/Preconditions/RequireBotPermissionAttribute.cs]: Can add more properties to set, but I wasn't sure about doing that.

File: src/Discord.Net.Commands/Attributes/Preconditions/RequireOwnerAttribute.cs [hunk 2]
```diff
@@ -10,17 +10,19 @@ namespace Discord.Commands
...
             {
                 case TokenType.Bot:
                     var application = await context.Client.GetApplicationInfoAsync();
                     if (context.User.Id != application.Owner.Id)
-                        return PreconditionResult.FromError("Command can only be run by the owner of the bot");
+                        return PreconditionResult.FromError(ErrorMessage ?? "Command can only be run by the owner of the bot");
                     return PreconditionResult.FromSuccess();
                 default:
-                    return PreconditionResult.FromError($"{nameof(RequireOwnerAttribute)} is not supported by this {nameof(TokenType)}.");                    
+                    return PreconditionResult.FromError($"{nameof(RequireOwnerAttribute)} is not supported by this {nameof(TokenType)}.");
```
Comment (k-boyle) [inline src/Discord.Net.Commands/Attributes/Preconditions/RequireOwnerAttribute.cs:50]: Not sure if this one also needs its own custom message or not considering the other token types.

Comment (Joe4evr) [reply src/Discord.Net.Commands/Attributes/Preconditions/RequireOwnerAttribute.cs:50]: I don't think we want to encourage creation of non-Bot applications.

**Review by FiniteReality — COMMENTED** [review]
I think this would make more sense as a method which returns a string, and takes an object[] for formatting parameters. That way, we can pass in contextual data to the string too.

File: src/Discord.Net.Commands/Attributes/Preconditions/RequireUserPermissionAttribute.cs [hunk 3]
```diff
@@ -11,6 +11,8 @@ public class RequireUserPermissionAttribute : PreconditionAttribute
     {
         public GuildPermission? GuildPermission { get; }
         public ChannelPermission? ChannelPermission { get; }
+        public override string ErrorMessage { get; set; }
+        public string NotAGuild { get; set; }
```
Comment (Chris-Johnston) [inline src/Discord.Net.Commands/Attributes/Preconditions/RequireUserPermissionAttribute.cs]: I think that `NotAGuildMessage` or `NotAGuildErrorMessage` would be a more fitting name.

**Pull request merged**